File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/visitor/hosting/DatabaseService.java
package com.bigboxcode.designpattern.visitor.hosting;

public class DatabaseService implements Service {

    private final double price = 100.00;
    private final double backPrice = 30.00;
    private int quantity;
    private boolean backupEnabled = false;

    public DatabaseService(int quantity) {
        this.quantity = quantity;
    }

    public DatabaseService(int quantity, boolean backupEnabled) {
        this.quantity = quantity;
        this.backupEnabled = backupEnabled;
    }

    public double getPrice() {
        return price;
    }

    public int getQuantity() {
        return quantity;
    }

    public double getBackPrice() {
        return backPrice;
    }

    public boolean isBackupEnabled() {
        return backupEnabled;
    }

    @Override
    public double accept(HostingCalculatorVisitor hostingCalculatorVisitor) {
        return hostingCalculatorVisitor.visit(this);
    }
}

File: design-patterns/ts/visitor/hosting-cost/database-service.ts
// database-service.ts

import HostingCalculatorVisitor from "./hosting-calculator-visitor";
import Service from "./service";


class DatabaseService implements Service {

    private readonly price: number = 100.00;
    private readonly backPrice: number = 30.00;
    private quantity: number;
    private backupEnabled: boolean = false;

    constructor(quantity: number, backupEnabled: boolean = false) {
        this.quantity = quantity;
        this.backupEnabled = backupEnabled;
    }

    getPrice(): number {
        return this.price;
    }

    getQuantity(): number {
        return this.quantity;
    }

    getBackPrice(): number {
        return this.backPrice;
    }

    isBackupEnabled(): boolean {
        return this.backupEnabled;
    }

    accept(hostingCalculatorVisitor: HostingCalculatorVisitor): number {
        return hostingCalculatorVisitor.visit(this);
    }
}

export default DatabaseService;